torch/distributions/relaxed_categorical.py

```diff
@@ -0,0 +1,110 @@
+import torch
+from torch.autograd import Variable
+from torch.nn.functional import log_softmax
+from torch.distributions import constraints
+from torch.distributions.utils import clamp_probs, broadcast_all
+from torch.distributions.categorical import Categorical
+from torch.distributions.distribution import Distribution
+from torch.distributions.transformed_distribution import TransformedDistribution
+from torch.distributions.transforms import ExpTransform
+
+
+def _log_softmax(x, *args, **kwargs):
+    if isinstance(x, torch.Tensor):
+        return log_softmax(Variable(x), *args, **kwargs).data
+    return log_softmax(x, *args, **kwargs)
+
+
+class ExpRelaxedOneHotCategorical(Distribution):
+    r"""
+    Creates a relaxed one-hot categorical distribution parameterized by `probs` and `temperature`.
+    Returns a log of a point in the simplex. Based on the interface to OneHotCategorical.
+
+    Implementation based on [1].
+
+    See also: :func:`torch.distributions.OneHotCategorical`
+
+    Example::
+
+        >>> m = OneHotCategorical(torch.Tensor([ 0.25, 0.25, 0.25, 0.25 ]))
+        >>> m.sample()  # equal probability of 0, 1, 2, 3
+         0
+         0
+         1
+         0
+        [torch.FloatTensor of size 4]
+
+    Args:
+        probs (Tensor or Variable): event probabilities
+
+    [1] The Concrete Distribution: A Continuous Relaxation of Discrete Random Variables
+    (Maddison et al, 2017)
+
+    [2] Categorical Reparametrization with Gumbel-Softmax
+    (Jang et al, 2017)
+    """
+    params = {'probs': constraints.simplex}
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, temperature, probs=None, logits=None):
+        self._categorical = Categorical(probs, logits)
+        self.temperature = temperature
+        batch_shape = self._categorical.batch_shape
+        event_shape = self._categorical.param_shape[-1:]
+        super(ExpRelaxedOneHotCategorical, self).__init__(batch_shape, event_shape)
+
+    def _new(self, *args, **kwargs):
+        return self._categorical._new(*args, **kwargs)
+
+    @property
+    def param_shape(self):
+        return self._categorical.param_shape
+
+    @property
+    def logits(self):
+        return self._categorical.logits
+
+    @property
+    def probs(self):
+        return self._categorical.probs
+
+    def rsample(self, sample_shape=torch.Size()):
+        sample_shape = torch.Size(sample_shape)
+        uniforms = clamp_probs(self.logits.new(self._extended_shape(sample_shape)).uniform_())
+        gumbels = uniforms.log().mul(-1).log().mul(-1)
```

Comment:
Is `__neg__` cheaper or more stable than `.mul(-1)`?

Reply:
Can I call it via `-(-(uniforms.log()).log())`?

Reply:
Yes, I just assume it's a tiny bit cheaper.